pkg/engine/internal/scheduler/scheduler.go

```diff
@@ -561,145 +632,134 @@ func (s *Scheduler) RemoveStreams(ctx context.Context, streams ...*workflow.Stre
+// findTasks gets a list of [task] from workflow tasks. Returns an error if any
+// of the tasks weren't recognized.
+func (s *Scheduler) findTasks(tasks []*workflow.Task) ([]*task, error) {
+	s.resourcesMut.RLock()
+	defer s.resourcesMut.RUnlock()
 
-	s.resourcesMut.Lock()
-	defer s.resourcesMut.Unlock()
+	res := make([]*task, 0, len(tasks))
 
 	var errs []error
 
-NextTask:
-	for _, taskToStart := range tasks {
-		if _, exist := s.tasks[taskToStart.ULID]; exist {
-			errs = append(errs, fmt.Errorf("task %s already exists", taskToStart.ULID))
-			continue
+	for _, task := range tasks {
+		if t, ok := s.tasks[task.ULID]; ok {
+			res = append(res, t)
+		} else {
+			errs = append(errs, fmt.Errorf("task %s not found", task.ULID))
 		}
+	}
 
-		for _, neededStreams := range taskToStart.Sources {
-			for _, neededStream := range neededStreams {
-				sourceStream, exist := s.streams[neededStream.ULID]
-				if !exist {
-					errs = append(errs, fmt.Errorf("source stream %s not found", neededStream))
-					continue NextTask
-				} else if err := sourceStream.setTaskReceiver(taskToStart.ULID); err != nil {
-					errs = append(errs, err)
-					continue NextTask
-				}
-			}
-		}
-		for _, neededStreams := range taskToStart.Sinks {
-			for _, neededStream := range neededStreams {
-				sinkStream, exist := s.streams[neededStream.ULID]
-				if !exist {
-					errs = append(errs, fmt.Errorf("sink stream %s not found", neededStream))
-					continue NextTask
-				} else if err := sinkStream.setTaskSender(taskToStart.ULID); err != nil {
-					errs = append(errs, err)
-					continue NextTask
-				}
-			}
-		}
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+	return res, nil
+}
 
-		newTask := &task{
-			inner:   taskToStart,
-			handler: handler,
+func (s *Scheduler) enqueueTasks(tasks []*task) error {
+	s.assignMut.Lock()
+	defer s.assignMut.Unlock()
+
+	var errs []error
 
-			// We initialize the status as Pending, which is an implicit
-			// transition from the default Created state.
-			status: workflow.TaskStatus{State: workflow.TaskStatePending},
+	for _, task := range tasks {
+		// Only allow to enqueue tasks in the initial state (created). This
+		// prevents tasks from accidentally being run multiple times.
+		if got, want := task.status.State, workflow.TaskStateCreated; got != want {
+			errs = append(errs, fmt.Errorf("task %s is in state %s, not %s", task.inner.ULID, got, want))
+			continue
 		}
-		s.tasks[taskToStart.ULID] = newTask
 
-		newTasks = append(newTasks, newTask)
+		s.taskQueue = append(s.taskQueue, task)
+	}
 
-		// Inform the owner about the state change from Created to Pending.
-		n.AddTaskEvent(taskNotification{
-			Handler:   handler,
-			Task:      taskToStart,
-			NewStatus: newTask.status,
-		})
+	if len(s.readyWorkers) > 0 && len(s.taskQueue) > 0 {
+		nudgeSemaphore(s.assignSema)
 	}
 
 	if len(errs) == 0 {
-		return newTasks, nil
+		return nil
 	}
-	return newTasks, errors.Join(errs...)
+	return errors.Join(errs...)
```

Comment:
nit: I think we should use a consistent way to return nil or err

```suggestion
if len(errs) > 0 {
		return errors.Join(errs...)
}
	return nil
```